packages/table/src/table.tsx

```diff
@@ -685,29 +687,49 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
     }
 
     protected validateProps(props: ITableProps & { children: React.ReactNode }) {
-        const { children, numFrozenColumns, numFrozenRows, numRows } = props;
+        const { children, columnWidths, numFrozenColumns, numFrozenRows, numRows, rowHeights } = props;
         const numColumns = React.Children.count(children);
 
+        // do cheap error-checking first.
+        if (numRows != null && numRows < 0) {
+            throw new Error(Errors.TABLE_NUM_ROWS_NEGATIVE);
+        }
+        if (numFrozenRows != null && numFrozenRows < 0) {
+            throw new Error(Errors.TABLE_NUM_FROZEN_ROWS_NEGATIVE);
+        }
+        if (numFrozenColumns != null && numFrozenColumns < 0) {
+            throw new Error(Errors.TABLE_NUM_FROZEN_COLUMNS_NEGATIVE);
+        }
+        if (numRows != null && rowHeights != null && rowHeights.length !== numRows) {
+            throw new Error(Errors.TABLE_NUM_ROWS_ROW_HEIGHTS_MISMATCH);
+        }
+        if (numColumns != null && columnWidths != null && columnWidths.length !== numColumns) {
+            throw new Error(Errors.TABLE_NUM_COLUMNS_COLUMN_WIDTHS_MISMATCH);
+        }
         React.Children.forEach(children, (child: React.ReactElement<any>) => {
             // save as a variable so that union type narrowing works
             const childType = child.type;
 
-            if (typeof childType === "string") {
-                console.warn(Errors.TABLE_NON_COLUMN_CHILDREN_WARNING);
+            // the second part of this conditional will never be true, but it
+            // informs the TS compiler that we won't be invoking
+            // childType.prototype on a "string" element.
+            if (typeof child === "string" || typeof childType === "string") {
+                throw new Error(Errors.TABLE_NON_COLUMN_CHILDREN_WARNING);
             } else {
                 const isColumn = childType.prototype === Column.prototype || Column.prototype.isPrototypeOf(childType);
                 if (!isColumn) {
-                    console.warn(Errors.TABLE_NON_COLUMN_CHILDREN_WARNING);
+                    throw new Error(Errors.TABLE_NON_COLUMN_CHILDREN_WARNING);
```

Comment:
Is there ever a case where we'd want to let people wrap child `<Column>`s with a thing? If not, would prefer this to be an `Error`.

Reply:
I think you're probably right, @adidahiya might know if the other ever comes up? Not sure.

Reply:
do we really need to validate the react children? what does this buy us? what's the harm in being lenient? please keep https://github.com/palantir/blueprint/issues/1108 in mind; enforcing constraints on React children in typescript usually gets in the way more often than it helps.

Reply:
`table.tsx` accesses `props.children` expecting it to be an array of `<Column>s` (e.g. in `Table.getColumnProps`). If that accessing fails, the error message is somewhat convoluted; throwing a clear error from `validateProps` would be more helpful.

Only reservation is that maybe there are cases when people would want to compose a basic `<Column>` into something else they'd want to provide as a direct child of `Table`? In that case, a duck-typing approach would seem preferable (i.e. as long as the child in question supports the `Column` API, we don't care).

I've never heard of anyone doing the above customization, so I'm including the error check as a result of my first point. ^ Thoughts?

Reply:
for now I'm fine with enforcing `Column` to be a direct child of `Table` -- as you said we haven't seen the wrapping component use case come up. makes sense to have a better error message and halt execution since the table won't render correctly. let's get to #1108 soon for 2.0.

Reply:
👍